phoenicis-engines/src/main/java/org/phoenicis/engines/EngineSettingsManager.java

```diff
@@ -19,98 +19,128 @@
  */
 public class EngineSettingsManager {
-    private final ScriptInterpreter scriptInterpreter;
+    private final PhoenicisScriptEngineFactory phoenicisScriptEngineFactory;
+    private final ExecutorService executorService;
 
     /**
-     * constructor
-     * @param scriptInterpreter
+     * Constructor
+     *
+     * @param phoenicisScriptEngineFactory The used script engine factory
+     * @param executorService The executor service to allow for parallelization
      */
-    public EngineSettingsManager(ScriptInterpreter scriptInterpreter) {
-        this.scriptInterpreter = scriptInterpreter;
-    }
+    public EngineSettingsManager(PhoenicisScriptEngineFactory phoenicisScriptEngineFactory,
+            ExecutorService executorService) {
+        super();
 
-    /**
-     * fetches the required setting
-     * @param engineId setting ID (e.g. "glsl")
-     * @param settingId setting ID (e.g. "glsl")
-     * @param doneCallback callback which will be executed with the fetched setting
-     * @param errorCallback callback which will be executed if an error occurs
-     */
-    public void getSetting(String engineId, String settingId, Consumer<EngineSetting> doneCallback,
-            Consumer<Exception> errorCallback) {
-        final InteractiveScriptSession interactiveScriptSession = scriptInterpreter.createInteractiveSession();
-
-        interactiveScriptSession.eval(
-                "include(\"engines." + engineId + ".settings." + settingId + "\"); new Setting();",
-                output -> {
-                    final EngineSetting setting = ((Value) output).as(EngineSetting.class);
-                    doneCallback.accept(setting);
-                }, errorCallback);
+        this.phoenicisScriptEngineFactory = phoenicisScriptEngineFactory;
+        this.executorService = executorService;
     }
 
     /**
-     * fetches the available engine settings
-     * @param repositoryDTO
-     * @param callback
-     * @param errorCallback callback which will be executed if an error occurs
+     * Fetches the available engine settings
+     *
+     * @param repositoryDTO The repository containing the engine settings
+     * @param callback The callback which recieves the found engine settings
+     * @param errorCallback The callback which will be executed if an error occurs
      */
     public void fetchAvailableEngineSettings(RepositoryDTO repositoryDTO,
             Consumer<Map<String, List<EngineSetting>>> callback, Consumer<Exception> errorCallback) {
-        final InteractiveScriptSession interactiveScriptSession = scriptInterpreter.createInteractiveSession();
+        executorService.execute(() -> {
+            final List<SettingConfig> configurations = fetchSettingConfigurations(repositoryDTO);
+
+            // the script engine needs to be created inside the correct thread otherwise GraalJS throwa an error
+            final PhoenicisScriptEngine phoenicisScriptEngine = phoenicisScriptEngineFactory.createEngine();
```

Comment:
Do we have this issue also in other places?

Reply:
I don't think so, but this the only place where I directly make use of the PhoenicisScriptEngine manager. Everything other place where we execute scripts uses a "managed" environment. I believe that this managed environment uses a single thread to both create the script engine and then to execute the script.